pkg/controller/storagecluster/placement_test.go

```diff
@@ -13,108 +13,105 @@ import (
-								Operator: corev1.NodeSelectorOpExists,
-							},
-						},
-					},
+var workerNodeSelector = corev1.NodeSelector{
+	NodeSelectorTerms: []corev1.NodeSelectorTerm{
+		corev1.NodeSelectorTerm{
+			MatchExpressions: []corev1.NodeSelectorRequirement{
+				corev1.NodeSelectorRequirement{
+					Key:      WorkerAffinityKey,
+					Operator: corev1.NodeSelectorOpExists,
 				},
 			},
 		},
-		Tolerations: []corev1.Toleration{
-			corev1.Toleration{
-				Key:      defaults.NodeTolerationKey,
-				Operator: corev1.TolerationOpEqual,
-				Value:    "true",
-				Effect:   corev1.TaintEffectNoSchedule,
+	},
+}
+
+var emptyLabelSelector = metav1.LabelSelector{
+	MatchExpressions: []metav1.LabelSelectorRequirement{},
+}
+var emptyNodeAffinity = corev1.NodeAffinity{
+	RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{
+		NodeSelectorTerms: []corev1.NodeSelectorTerm{
+			corev1.NodeSelectorTerm{
+				MatchExpressions: []corev1.NodeSelectorRequirement{},
 			},
 		},
 	},
 }
+var emptyPlacements = map[rookv1.KeyType]rookv1.Placement{
+	"all": rookv1.Placement{},
+}
 
 func TestGetPlacement(t *testing.T) {
-	// Case 1: Defaults are preserved i.e no placement and no label selector
-	sc := &ocsv1.StorageCluster{}
-	mockStorageCluster.DeepCopyInto(sc)
-	assert.Equal(t, defaultLabelPlacement["all"], getPlacement(sc, "all"))
-
-	// Case 2: The configured Placements override the defaults
-	sc = &ocsv1.StorageCluster{}
-	mockStorageCluster.DeepCopyInto(sc)
-	sc.Spec.Placement = mockPlacements
-	assert.Equal(t, defaultLabelPlacement["all"], getPlacement(sc, "all"))
-
-	// Case 3: LabelSelector to modify the default placements correctly
-	sc = &ocsv1.StorageCluster{}
-	mockStorageCluster.DeepCopyInto(sc)
-	sc.Spec.LabelSelector = &metav1.LabelSelector{
-		MatchExpressions: []metav1.LabelSelectorRequirement{
-			metav1.LabelSelectorRequirement{
-				Key:      WorkerAffinityKey,
-				Operator: metav1.LabelSelectorOpExists,
+	cases := []struct {
+		placements        rookv1.PlacementSpec
+		labelSelector     *metav1.LabelSelector
+		expectedPlacement rookv1.Placement
+	}{
+		{
+			// Case 1: Defaults are preserved i.e no placement and no label selector
+			placements:    rookv1.PlacementSpec{},
+			labelSelector: nil,
+			expectedPlacement: rookv1.Placement{
+				NodeAffinity: defaults.DefaultNodeAffinity,
+				Tolerations:  defaults.DaemonPlacements["all"].Tolerations,
 			},
 		},
-	}
-	assert.Equal(t, mockPlacements["all"], getPlacement(sc, "all"))
-
-	// Case 4: The LabelSelector modifies the configured Placements correctly
-	sc = &ocsv1.StorageCluster{}
-	mockStorageCluster.DeepCopyInto(sc)
-	sc.Spec.Placement = mockPlacements
-	sc.Spec.LabelSelector = &metav1.LabelSelector{
-		MatchExpressions: []metav1.LabelSelectorRequirement{
-			metav1.LabelSelectorRequirement{
-				Key:      MasterAffinityKey,
-				Operator: metav1.LabelSelectorOpExists,
+		{
+			// Case 2: The configured Placements override the defaults
```

Comment:
I still see default under expectedPlacement?? and, the configured placement is actually empty.

Reply:
I think placement has been provided but it just maps `"all"` to an empty placement, which is overridden by the default placement?

Reply:
Correct, the LabelSelector always modifies the NodeAffinity of all Placements. Only when the LabelSelector is defined as empty (Case 5) will the NodeAffinity not be set. This gives preference for the behavior of ALWAYS tying our Pods to storage nodes, with the requirement to take a manual step to disable it.